File: domain/src/test/java/org/cryptomator/domain/usecases/cloud/DataSourceCapturingAnswer.java
package org.cryptomator.domain.usecases.cloud;

import org.mockito.invocation.InvocationOnMock;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

class DataSourceCapturingAnswer<T> implements org.mockito.stubbing.Answer<T> {

	private final T result;
	private final int argIndex;
	private ByteArrayOutputStream out;

	DataSourceCapturingAnswer(T result, int argIndex) {
		this.result = result;
		this.argIndex = argIndex;
	}

	@Override
	public T answer(InvocationOnMock invocation) throws Throwable {
		InputStream in = ((DataSource) invocation.getArguments()[argIndex]).open(null);
		out = new ByteArrayOutputStream();
		copy(in, out);
		return result;
	}

	private void copy(InputStream in, ByteArrayOutputStream out) {
		byte[] buffer = new byte[4096];
		int read;
		try {
			while ((read = in.read(buffer)) != -1) {
				out.write(buffer, 0, read);
			}
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	public byte[] toByteArray() {
		return out.toByteArray();
	}

}

File: domain/src/test/java/org/cryptomator/domain/usecases/cloud/DataSourceCapturingAnswer.kt
package org.cryptomator.domain.usecases.cloud

import android.content.Context
import org.cryptomator.domain.exception.FatalBackendException
import org.mockito.invocation.InvocationOnMock
import org.mockito.kotlin.mock
import org.mockito.stubbing.Answer
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream

internal class DataSourceCapturingAnswer<T>(private val result: T, private val argIndex: Int) : Answer<T> {

	private lateinit var out: ByteArrayOutputStream

	private val context : Context = mock()

	@Throws(Throwable::class)
	override fun answer(invocation: InvocationOnMock): T {
		val inputStream = (invocation.arguments[argIndex] as DataSource).open(context)!!
		out = ByteArrayOutputStream()
		copy(inputStream, out)
		return result
	}

	private fun copy(inputStream: InputStream, out: ByteArrayOutputStream) {
		val buffer = ByteArray(4096)
		var read: Int
		try {
			while (inputStream.read(buffer).also { read = it } != -1) {
				out.write(buffer, 0, read)
			}
		} catch (e: IOException) {
			throw FatalBackendException(e)
		}
	}

	fun toByteArray(): ByteArray {
		return out.toByteArray()
	}
}
